In [47]:
import pandas as pd
import datetime

# First, read in station data

In [2]:
# Station metadata
stn_ids = pd.read_fwf('http://noaa-ghcn-pds.s3.amazonaws.com/ghcnd-stations.txt', header=None, infer_nrows=1000)
stn_ids.columns = ['ID','LAT','LON','ELEV','UKN','NAME','GSN','WBAN']
stn_ids

ID      LAT      LON    ELEV  UKN                   NAME  \
0       ACW00011604  17.1167 -61.7833    10.1  NaN  ST JOHNS COOLIDGE FLD   
1       ACW00011647  17.1333 -61.7833    19.2  NaN               ST JOHNS   
2       AE000041196  25.3330  55.5170    34.0  NaN    SHARJAH INTER. AIRP   
3       AEM00041194  25.2550  55.3640    10.4  NaN             DUBAI INTL   
4       AEM00041217  24.4330  54.6510    26.8  NaN         ABU DHABI INTL   
...             ...      ...      ...     ...  ...                    ...   
129653  ZI000067969 -21.0500  29.3670   861.0  NaN         WEST NICHOLSON   
129654  ZI000067975 -20.0670  30.8670  1095.0  NaN               MASVINGO   
129655  ZI000067977 -21.0170  31.5830   430.0  NaN          BUFFALO RANGE   
129656  ZI000067983 -20.2000  32.6160  1132.0  NaN               CHIPINGE   
129657  ZI000067991 -22.2170  30.0000   457.0  NaN             BEITBRIDGE   

        GSN     WBAN  
0       NaN      NaN  
1       NaN      NaN  
2       GSN  41196.0  
3       NaN  41194.0  
4       NaN  41217.0  
...     ...      ...  
129653  NaN  67969.0  
129654  NaN  67975.0  
129655  NaN  67977.0  
129656  GSN  67983.0  
129657  NaN  67991.0  

[129658 rows x 8 columns]

In [3]:
# Station periods of record
periods = pd.read_fwf('http://noaa-ghcn-pds.s3.amazonaws.com/ghcnd-inventory.txt', header=None, infer_nrows=1000)
periods.columns = ['ID','LAT','LON','ELEM','TiMIN','TiMAX']
periods

ID      LAT      LON  ELEM  TiMIN  TiMAX
0       ACW00011604  17.1167 -61.7833  TMAX   1949   1949
1       ACW00011604  17.1167 -61.7833  TMIN   1949   1949
2       ACW00011604  17.1167 -61.7833  PRCP   1949   1949
3       ACW00011604  17.1167 -61.7833  SNOW   1949   1949
4       ACW00011604  17.1167 -61.7833  SNWD   1949   1949
...             ...      ...      ...   ...    ...    ...
767508  ZI000067983 -20.2000  32.6160  PRCP   1951   2025
767509  ZI000067983 -20.2000  32.6160  TAVG   1962   2025
767510  ZI000067991 -22.2170  30.0000  TMAX   1951   1990
767511  ZI000067991 -22.2170  30.0000  TMIN   1951   1990
767512  ZI000067991 -22.2170  30.0000  PRCP   1951   1990

[767513 rows x 6 columns]

In [5]:
# Merge the two DataFrames
merged_stns = pd.merge(stn_ids,periods,how='left',left_on='ID',right_on='ID')
merged_stns

ID    LAT_x    LON_x    ELEV  UKN                   NAME  \
0       ACW00011604  17.1167 -61.7833    10.1  NaN  ST JOHNS COOLIDGE FLD   
1       ACW00011604  17.1167 -61.7833    10.1  NaN  ST JOHNS COOLIDGE FLD   
2       ACW00011604  17.1167 -61.7833    10.1  NaN  ST JOHNS COOLIDGE FLD   
3       ACW00011604  17.1167 -61.7833    10.1  NaN  ST JOHNS COOLIDGE FLD   
4       ACW00011604  17.1167 -61.7833    10.1  NaN  ST JOHNS COOLIDGE FLD   
...             ...      ...      ...     ...  ...                    ...   
767572  ZI000067983 -20.2000  32.6160  1132.0  NaN               CHIPINGE   
767573  ZI000067983 -20.2000  32.6160  1132.0  NaN               CHIPINGE   
767574  ZI000067991 -22.2170  30.0000   457.0  NaN             BEITBRIDGE   
767575  ZI000067991 -22.2170  30.0000   457.0  NaN             BEITBRIDGE   
767576  ZI000067991 -22.2170  30.0000   457.0  NaN             BEITBRIDGE   

        GSN     WBAN    LAT_y    LON_y  ELEM   TiMIN   TiMAX  
0       NaN      NaN  17.1167 -61.7833  TMAX  1949.0  1949.0  
1       NaN      NaN  17.1167 -61.7833  TMIN  1949.0  1949.0  
2       NaN      NaN  17.1167 -61.7833  PRCP  1949.0  1949.0  
3       NaN      NaN  17.1167 -61.7833  SNOW  1949.0  1949.0  
4       NaN      NaN  17.1167 -61.7833  SNWD  1949.0  1949.0  
...     ...      ...      ...      ...   ...     ...     ...  
767572  GSN  67983.0 -20.2000  32.6160  PRCP  1951.0  2025.0  
767573  GSN  67983.0 -20.2000  32.6160  TAVG  1962.0  2025.0  
767574  NaN  67991.0 -22.2170  30.0000  TMAX  1951.0  1990.0  
767575  NaN  67991.0 -22.2170  30.0000  TMIN  1951.0  1990.0  
767576  NaN  67991.0 -22.2170  30.0000  PRCP  1951.0  1990.0  

[767577 rows x 13 columns]

In [6]:
# Grab all stations that have TMIN data from at least 1990 to 2020
merged_stns = merged_stns[(merged_stns['ELEM'] == 'TMIN') & (merged_stns['TiMAX'] >= 2020) & (merged_stns['TiMIN'] <= 1990)]
merged_stns

ID   LAT_x   LON_x    ELEV  UKN                 NAME  GSN  \
19      AE000041196  25.333  55.517    34.0  NaN  SHARJAH INTER. AIRP  GSN   
23      AEM00041194  25.255  55.364    10.4  NaN           DUBAI INTL  NaN   
27      AEM00041217  24.433  54.651    26.8  NaN       ABU DHABI INTL  NaN   
40      AFM00040938  34.210  62.228   977.2  NaN                HERAT  NaN   
45      AFM00040948  34.566  69.212  1791.3  NaN           KABUL INTL  NaN   
...             ...     ...     ...     ...  ...                  ...  ...   
767505  ZA000067743 -17.817  25.817   986.0  NaN          LIVINGSTONE  GSN   
767510  ZAM00067663 -14.450  28.467  1207.0  NaN       KABWE/MILLIKEN  NaN   
767522  ZI000067775 -17.917  31.133  1480.0  NaN     HARARE (KUTSAGA)  GSN   
767563  ZI000067975 -20.067  30.867  1095.0  NaN             MASVINGO  NaN   
767571  ZI000067983 -20.200  32.616  1132.0  NaN             CHIPINGE  GSN   

           WBAN   LAT_y   LON_y  ELEM   TiMIN   TiMAX  
19      41196.0  25.333  55.517  TMIN  1944.0  2025.0  
23      41194.0  25.255  55.364  TMIN  1983.0  2025.0  
27      41217.0  24.433  54.651  TMIN  1983.0  2025.0  
40      40938.0  34.210  62.228  TMIN  1973.0  2020.0  
45      40948.0  34.566  69.212  TMIN  1967.0  2021.0  
...         ...     ...     ...   ...     ...     ...  
767505  67743.0 -17.817  25.817  TMIN  1973.0  2025.0  
767510  67663.0 -14.450  28.467  TMIN  1973.0  2025.0  
767522  67775.0 -17.917  31.133  TMIN  1956.0  2025.0  
767563  67975.0 -20.067  30.867  TMIN  1951.0  2025.0  
767571  67983.0 -20.200  32.616  TMIN  1951.0  2025.0  

[9496 rows x 13 columns]

In [8]:
station_id = 'USC00087205'
# Make sure our station exists and its period of record is sufficient
merged_stns.loc[merged_stns['ID'] == station_id]

ID    LAT_x    LON_x  ELEV UKN        NAME  GSN  WBAN  \
469182  USC00087205  28.0208 -82.1392  33.2  FL  PLANT CITY  NaN   NaN   

          LAT_y    LON_y  ELEM   TiMIN   TiMAX  
469182  28.0208 -82.1392  TMIN  1892.0  2025.0

In [10]:
# Load the station data from S3 bucket
df = pd.read_csv(
    f"s3://noaa-ghcn-pds/csv/by_station/{station_id}.csv",
    storage_options={"anon": True},  # passed to `s3fs.S3FileSystem`
    dtype={'Q_FLAG': 'object', 'M_FLAG': 'object'},
    parse_dates=['DATE']
).set_index('DATE')

df

C:\Users\domenicbrooks\AppData\Local\Temp\ipykernel_23976\2101759641.py:2: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(


ID ELEMENT  DATA_VALUE M_FLAG Q_FLAG S_FLAG  OBS_TIME
DATE                                                                      
1892-09-01  USC00087205    TMAX         322    NaN    NaN      6       NaN
1892-09-02  USC00087205    TMAX         317    NaN    NaN      6       NaN
1892-09-03  USC00087205    TMAX         317    NaN    NaN      6       NaN
1892-09-04  USC00087205    TMAX         322    NaN    NaN      6       NaN
1892-09-05  USC00087205    TMAX         333    NaN    NaN      6       NaN
...                 ...     ...         ...    ...    ...    ...       ...
2025-12-10  USC00087205    PRCP           0    NaN    NaN      H    1600.0
2025-12-11  USC00087205    PRCP           0    NaN    NaN      H    1600.0
2025-12-12  USC00087205    PRCP           0    NaN    NaN      H    1600.0
2025-12-13  USC00087205    PRCP           0    NaN    NaN      H    1600.0
2025-12-14  USC00087205    PRCP           0    NaN    NaN      H    1600.0

[224247 rows x 7 columns]

In [ ]:
# Grab TMIN data since we are concerned with freezing days
df_tmin = df.loc[df['ELEMENT'] == 'TMIN']
df_tmin['DATA_VALUE'] = df_tmin['DATA_VALUE'] / 10.0  # Convert to deg C
df_tmin

C:\Users\domenicbrooks\AppData\Local\Temp\ipykernel_23976\859045960.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_tmin['DATA_VALUE'] = df_tmin['DATA_VALUE'] / 10.0  # Convert to deg C


ID ELEMENT  DATA_VALUE M_FLAG Q_FLAG S_FLAG  OBS_TIME
DATE                                                                      
1892-09-01  USC00087205    TMIN        20.6    NaN    NaN      6       NaN
1892-09-02  USC00087205    TMIN        20.6    NaN    NaN      6       NaN
1892-09-03  USC00087205    TMIN        21.1    NaN    NaN      6       NaN
1892-09-04  USC00087205    TMIN        21.7    NaN    NaN      6       NaN
1892-09-05  USC00087205    TMIN        21.1    NaN    NaN      6       NaN
...                 ...     ...         ...    ...    ...    ...       ...
2025-12-10  USC00087205    TMIN        12.8    NaN    NaN      H    1600.0
2025-12-11  USC00087205    TMIN        10.0    NaN    NaN      H    1600.0
2025-12-12  USC00087205    TMIN         6.7    NaN    NaN      H    1600.0
2025-12-13  USC00087205    TMIN        10.0    NaN    NaN      H    1600.0
2025-12-14  USC00087205    TMIN        15.6    NaN    NaN      H    1600.0

[45420 rows x 7 columns]

### Now that we have the min temp data from Plant City, lets move onto question 1.

Goal: Mean number of days per month (for October-January) for 1991-2020 period that were less than or equal to 32 deg F (frost) and 28 deg F (freeze).

In [ ]:
# Convert to deg F first
df_tmin_f = df_tmin.copy()
df_tmin_f['DATA_VALUE'] = ((df_tmin['DATA_VALUE'] * (9/5)) + 32) 
df_tmin_f

ID ELEMENT  DATA_VALUE M_FLAG Q_FLAG S_FLAG  OBS_TIME
DATE                                                                      
1892-09-01  USC00087205    TMIN       69.08    NaN    NaN      6       NaN
1892-09-02  USC00087205    TMIN       69.08    NaN    NaN      6       NaN
1892-09-03  USC00087205    TMIN       69.98    NaN    NaN      6       NaN
1892-09-04  USC00087205    TMIN       71.06    NaN    NaN      6       NaN
1892-09-05  USC00087205    TMIN       69.98    NaN    NaN      6       NaN
...                 ...     ...         ...    ...    ...    ...       ...
2025-12-10  USC00087205    TMIN       55.04    NaN    NaN      H    1600.0
2025-12-11  USC00087205    TMIN       50.00    NaN    NaN      H    1600.0
2025-12-12  USC00087205    TMIN       44.06    NaN    NaN      H    1600.0
2025-12-13  USC00087205    TMIN       50.00    NaN    NaN      H    1600.0
2025-12-14  USC00087205    TMIN       60.08    NaN    NaN      H    1600.0

[45420 rows x 7 columns]

In [30]:
# Select TMIN data for 1991-2020 and months October to January
df_tmin_f_period = df_tmin_f[(df_tmin_f.index.year >= 1991)
                              & (df_tmin_f.index.year <= 2020)
                              & (df_tmin_f.index.month.isin([10,11,12,1]))]

# Add month of year column as well
df_tmin_f_period['MONTH'] = df_tmin_f_period.index.month
df_tmin_f_period

C:\Users\domenicbrooks\AppData\Local\Temp\ipykernel_23976\1680339355.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_tmin_f_period['MONTH'] = df_tmin_f_period.index.month


ID ELEMENT  DATA_VALUE M_FLAG Q_FLAG S_FLAG  OBS_TIME  \
DATE                                                                         
1991-01-01  USC00087205    TMIN       62.06    NaN    NaN      0    1800.0   
1991-01-02  USC00087205    TMIN       64.94    NaN    NaN      0    1800.0   
1991-01-03  USC00087205    TMIN       64.04    NaN    NaN      0    1800.0   
1991-01-04  USC00087205    TMIN       64.04    NaN    NaN      0    1800.0   
1991-01-05  USC00087205    TMIN       62.96    NaN    NaN      0    1800.0   
...                 ...     ...         ...    ...    ...    ...       ...   
2020-12-27  USC00087205    TMIN       33.98    NaN    NaN      7    1600.0   
2020-12-28  USC00087205    TMIN       48.02    NaN    NaN      7    1600.0   
2020-12-29  USC00087205    TMIN       53.06    NaN    NaN      7    1600.0   
2020-12-30  USC00087205    TMIN       55.04    NaN    NaN      7    1600.0   
2020-12-31  USC00087205    TMIN       66.02    NaN    NaN      7    1600.0   

            MONTH  
DATE               
1991-01-01      1  
1991-01-02      1  
1991-01-03      1  
1991-01-04      1  
1991-01-05      1  
...           ...  
2020-12-27     12  
2020-12-28     12  
2020-12-29     12  
2020-12-30     12  
2020-12-31     12  

[3617 rows x 8 columns]

In [32]:
# Add columns to denote frost and freeze days
df_tmin_f_period['FROST'] = df_tmin_f_period['DATA_VALUE'] <= 32
df_tmin_f_period['FREEZE'] = df_tmin_f_period['DATA_VALUE'] <= 28

df_tmin_f_period

C:\Users\domenicbrooks\AppData\Local\Temp\ipykernel_23976\540180356.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_tmin_f_period['FROST'] = df_tmin_f_period['DATA_VALUE'] <= 32
C:\Users\domenicbrooks\AppData\Local\Temp\ipykernel_23976\540180356.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_tmin_f_period['FREEZE'] = df_tmin_f_period['DATA_VALUE'] <= 28


ID ELEMENT  DATA_VALUE M_FLAG Q_FLAG S_FLAG  OBS_TIME  \
DATE                                                                         
1991-01-01  USC00087205    TMIN       62.06    NaN    NaN      0    1800.0   
1991-01-02  USC00087205    TMIN       64.94    NaN    NaN      0    1800.0   
1991-01-03  USC00087205    TMIN       64.04    NaN    NaN      0    1800.0   
1991-01-04  USC00087205    TMIN       64.04    NaN    NaN      0    1800.0   
1991-01-05  USC00087205    TMIN       62.96    NaN    NaN      0    1800.0   
...                 ...     ...         ...    ...    ...    ...       ...   
2020-12-27  USC00087205    TMIN       33.98    NaN    NaN      7    1600.0   
2020-12-28  USC00087205    TMIN       48.02    NaN    NaN      7    1600.0   
2020-12-29  USC00087205    TMIN       53.06    NaN    NaN      7    1600.0   
2020-12-30  USC00087205    TMIN       55.04    NaN    NaN      7    1600.0   
2020-12-31  USC00087205    TMIN       66.02    NaN    NaN      7    1600.0   

            MONTH  FROST  FREEZE  
DATE                              
1991-01-01      1  False   False  
1991-01-02      1  False   False  
1991-01-03      1  False   False  
1991-01-04      1  False   False  
1991-01-05      1  False   False  
...           ...    ...     ...  
2020-12-27     12  False   False  
2020-12-28     12  False   False  
2020-12-29     12  False   False  
2020-12-30     12  False   False  
2020-12-31     12  False   False  

[3617 rows x 10 columns]

In [92]:
# Number of years in the period
num_years = len(df_tmin_f_period.index.year.unique().values)

In [93]:
# Group by month and sum the number of frost and freeze days over the 30-year period
df_mean_ff_days = df_tmin_f_period.groupby('MONTH').agg({'FROST':'sum','FREEZE':'sum'}) / num_years  # Divide by 30 years to get mean number of days per month
df_mean_ff_days

FROST    FREEZE
MONTH                    
1      1.866667  0.500000
10     0.000000  0.000000
11     0.033333  0.000000
12     0.600000  0.166667

In [98]:
print('Mean Number of Frost Days:\n'
      f'Jan: {df_mean_ff_days["FROST"].loc[1]:.2f}\n'
      f'Oct: {df_mean_ff_days["FROST"].loc[10]:.2f}\n'
      f'Nov: {df_mean_ff_days["FROST"].loc[11]:.2f}\n'
      f'Dec: {df_mean_ff_days["FROST"].loc[12]:.2f}\n')

print('Mean Number of Freeze Days:\n'
      f'Jan: {df_mean_ff_days["FREEZE"].loc[1]:.2f}\n'
      f'Oct: {df_mean_ff_days["FREEZE"].loc[10]:.2f}\n'
      f'Nov: {df_mean_ff_days["FREEZE"].loc[11]:.2f}\n'
      f'Dec: {df_mean_ff_days["FREEZE"].loc[12]:.2f}\n')

Mean Number of Frost Days:
Jan: 1.87
Oct: 0.00
Nov: 0.03
Dec: 0.60

Mean Number of Freeze Days:
Jan: 0.50
Oct: 0.00
Nov: 0.00
Dec: 0.17



### Part 2: ENSO patterns

Goal: Determine which ENSO region is best correlated (highest Pearson-correlation) with the number of days per month <28 def F (Freeze days)

In [ ]:
# Read in full ENSO dataset
full_enso = pd.read_csv('https://www.cpc.ncep.noaa.gov/data/indices/sstoi.indices',
                        sep=r'\s+', header=0)

# Rename month column for merging later
full_enso.rename(columns={'MON': 'MONTH'}, inplace=True)
full_enso.sort_values(['YR','MONTH'], inplace=True) # Sort just to be sure
full_enso

YR  MONTH  NINO1+2  ANOM  NINO3  ANOM.1  NINO4  ANOM.2  NINO3.4  ANOM.3
0    1982      1    24.28 -0.24  25.84    0.17  28.01   -0.21    26.65    0.08
1    1982      2    25.38 -0.72  26.26   -0.11  27.99   -0.11    26.54   -0.20
2    1982      3    25.22 -1.38  26.92   -0.25  28.18   -0.05    27.09   -0.14
3    1982      4    24.57 -1.16  27.52   -0.05  28.61    0.10    27.83    0.02
4    1982      5    24.00 -0.62  27.70    0.49  29.19    0.40    28.37    0.49
..    ...    ...      ...   ...    ...     ...    ...     ...      ...     ...
522  2025      7    22.29  0.46  25.92    0.04  28.84    0.05    27.24   -0.06
523  2025      8    21.09  0.23  24.97   -0.24  28.63   -0.06    26.58   -0.33
524  2025      9    20.40 -0.18  24.60   -0.41  28.41   -0.27    26.32   -0.44
525  2025     10    20.83 -0.04  24.74   -0.35  28.36   -0.33    26.29   -0.48
526  2025     11    21.28 -0.34  24.60   -0.60  28.15   -0.52    26.14   -0.68

[527 rows x 10 columns]

In [102]:
# Creat a new DataFrame with a Date index covering the full period of the ENSO dataset
enso_new=pd.DataFrame()
enso_new['Date']=pd.date_range(start=datetime.datetime(full_enso['YR'][0],1,1),end=datetime.datetime(full_enso['YR'].iloc[-1],12,1),freq="MS")

# Set Date as index and add MONTH column
enso_new = enso_new.set_index('Date').sort_index()
enso_new['MONTH'] = enso_new.index.month
enso_new = enso_new.iloc[:-1] # Drop last row since it goes beyond the ENSO data period

# Add ENSO anomaly columns to new DataFrame
enso_new['anom1+2'] = full_enso['ANOM'].values
enso_new['anom3'] = full_enso['ANOM.1'].values
enso_new['anom4'] = full_enso['ANOM.2'].values
enso_new['anom3.4'] = full_enso['ANOM.3'].values
enso_new


MONTH  anom1+2  anom3  anom4  anom3.4
Date                                             
1982-01-01      1    -0.24   0.17  -0.21     0.08
1982-02-01      2    -0.72  -0.11  -0.11    -0.20
1982-03-01      3    -1.38  -0.25  -0.05    -0.14
1982-04-01      4    -1.16  -0.05   0.10     0.02
1982-05-01      5    -0.62   0.49   0.40     0.49
...           ...      ...    ...    ...      ...
2025-07-01      7     0.46   0.04   0.05    -0.06
2025-08-01      8     0.23  -0.24  -0.06    -0.33
2025-09-01      9    -0.18  -0.41  -0.27    -0.44
2025-10-01     10    -0.04  -0.35  -0.33    -0.48
2025-11-01     11    -0.34  -0.60  -0.52    -0.68

[527 rows x 5 columns]

In [78]:
# First need to sum up freeze days by month and year
df_tmin_f_period['FREEZE_days_alt'] = df_tmin_f_period['DATA_VALUE'] < 28 # Only days <28F 
df_monthly_freeze_days = df_tmin_f_period.groupby(pd.Grouper(freq='MS')).agg({'FREEZE_days_alt':'sum'})
df_monthly_freeze_days

C:\Users\domenicbrooks\AppData\Local\Temp\ipykernel_23976\2350549940.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_tmin_f_period['FREEZE_days_alt'] = df_tmin_f_period['DATA_VALUE'] < 28 # Only days <28F


FREEZE_days_alt
DATE                       
1991-01-01                0
1991-02-01                0
1991-03-01                0
1991-04-01                0
1991-05-01                0
...                     ...
2020-08-01                0
2020-09-01                0
2020-10-01                0
2020-11-01                0
2020-12-01                0

[360 rows x 1 columns]

In [ ]:
# Now merge the two DataFrames on shared date index
df_merged = df_monthly_freeze_days.merge(enso_new, left_index=True, right_index=True, how='left')
df_merged = df_merged[df_merged.index.month.isin([10,11,12,1])] # Keep only Oct-Jan months
df_merged

FREEZE_days_alt  MONTH  anom1+2  anom3  anom4  anom3.4
DATE                                                              
1991-01-01                0      1    -0.78  -0.05   0.40     0.33
1991-10-01                0     10     0.22   0.27   0.63     0.64
1991-11-01                0     11     0.29   0.73   0.44     0.89
1991-12-01                0     12     0.29   1.03   0.77     1.50
1992-01-01                0      1     0.07   1.18   0.58     1.67
...                     ...    ...      ...    ...    ...      ...
2019-12-01                0     12     0.27   0.29   0.85     0.44
2020-01-01                0      1    -0.03   0.22   0.94     0.52
2020-10-01                0     10    -0.76  -0.90  -0.71    -1.13
2020-11-01                0     11    -0.67  -1.03  -0.70    -1.23
2020-12-01                0     12    -0.86  -0.70  -0.85    -0.99

[120 rows x 6 columns]

In [104]:
# Correlation between freeze days and each anomaly
print('Pearson correlations:\n')
for col in ['anom1+2', 'anom3', 'anom4', 'anom3.4']:
    corr = df_merged['FREEZE_days_alt'].corr(df_merged[col], method='pearson')
    print(f'{col}: {corr:.4f}')

Pearson correlations:

anom1+2: -0.1356
anom3: -0.1225
anom4: -0.1351
anom3.4: -0.1120


### It would appear as though all the ENSO regions are weakly negatively correlated with the monthly number of days <28 deg F for Oct-Jan, but the region with the highest magnitude correlation appears to be region 1+2, with -0.1356.